features/294-login-without-specifying-login-id-key/api.md

```diff
@@ -0,0 +1,370 @@
+// creates 3 loginIDs
+// 1. username: 'example',
+// 2. email: 'example@example.com'
+// 3. role-phone: { role: 'example , phone: '8912345' }
+skygear.auth.login(
+  {
+    role: 'example',
+    phone: '8912345',
+  }
+).then((user) => {
+  console.log(user);
+  /*
+  LoginID-Source: role-phone,
+  {
+      user_id: <id>,
+      login_ids: {
+        "role-phone": {
+          role: 'example',
+          phone: '8912345',
+        },
+        username: 'example',
+        email: 'example@example.com'
+      },
+      metadata: {
+        age: 18
+      }
+  }
+  */
+});
+// or
+skygear.auth.login(
+  'example'
+).then((user) => {
+  console.log(user);
+  /*
+  LoginID-Source: username,
+  {
+      user_id: <id>,
+        login_ids: {
+        "role-phone": {
+          role: 'example',
+          phone: '8912345',
+        },
+        username: 'example',
+        email: 'example@example.com'
+      },
+      metadata: {
+        age: 18
+      }
+  }
+  */
+});
+```
+
+### Code change
+
+1. [skygear-server] Remove `loginIDMetadataKeys` setting.
+2. [skygear-server] Change `_auth_provider_password.authData` to `text` type and add `_auth_provider_password.source` field.
+3. [skygear-JS-SDK] Update SDK APIs
+
+   | Old | New |
+   | -------- | -------- |
+   | `login(authData: dict, password: String)` | `login(loginID: String/dict, password: String)` |
+   | ` signup(authData: dict, password: String, data: dict)` | ` signup(loginIDs: dict, password: String, data: dict)` |
+   | `requestVerification(recordKey: String)` | `requestVerification(LoginIDSource: String, type: "email, phone")` |
+   | N/A | `createLoginID(loginIDSource: String, loginID: String/dict)`<br/>`updateLoginID(loginIDSource: String, loginID: String/dict)`<br/>`removeLoginID(loginIDSource: String)` |
+   | N/A | `createPasscode(loginIDSource: String, password: String, skip2FA: Boolean)`<br/>`updatePasscode(loginIDSource: String, password: String, skip2FA: Boolean`<br/>`deletePasscode(loginIDSource: String)` |
+4. [skygear-server] Update signup and login handler to handle the `loginID` concept.
+5. [skygear-server] Include `LoginID-Source` in response header.
+6. [skygear-server] Implement `requestVerification(LoginIDSource: String, type: "email"|"phone")`
+
+### SSO auto-link
+
+For SSO auto-link user feature, have following possible solutions:
+
+1. assume `loginID` is email.
+2. provide an extra auth-hook, which request to return `userID` by email.
+3. Check if there is email information in the metadata:
+   1. if there is a `sso_matching_emails`, all emails in the array of value will be matched against the SSO login.
+   2. if ther is a `email`, that email will be matched against the SSO login.
+
+Solution 3 may have uniqueness issue, i.e. two user could have identical email or overlapping email in `sso_matching_emails`.
+
+## Proposal 3
+
+Adjust `LOGIN_ID_METADATA_KEYS` definition, allows one level string `loginID` only.
+
+```
+LOGIN_SOURCES = ["username", "email"]
```

Comment:
Suggested to change it to `LOGIN_IDS_KEYS_WHITELIST`

Also, make it default as empty, and empty meant any loginIDs key could be used.

This is to avoid new users, say they want to try Skygear, and use `callsign` as the LoginID, but turns out cannot login and confused.

I think whitelist is a good balance for server side checking if it is needed.